src/awkward/operations/structure.py

```diff
@@ -1460,6 +1460,60 @@ def apply(layout):
         return out
 
 
+def unflatten(array, counts, highlevel=True):
+    """
+    Args:
+        array: Data to create an array with an additional level from.
+        counts (int or array): Number of elements the new level should have.
+        highlevel (bool): If True, return an #ak.Array; otherwise, return
+            a low-level #ak.layout.Content subclass.
+
+    Returns an array with an additional level of nesting. This is roughly the
+    inverse of #ak.flatten, where `counts` were obtained by #ak.num (both with
+    `axis=1`).
+    For example,
+
+        >>> original = ak.Array([[0, 1, 2], [], [3, 4], [5], [6, 7, 8, 9]])
+        >>> counts = ak.num(original)
+        >>> array = ak.flatten(original)
+        >>> counts
+        <Array [3, 0, 2, 1, 4] type='5 * int64'>
+        >>> array
+        <Array [0, 1, 2, 3, 4, 5, 6, 7, 8, 9] type='10 * int64'>
+        >>> ak.unflatten(array, counts)
+        <Array [[0, 1, 2], [], ... [5], [6, 7, 8, 9]] type='5 * var * int64'>
+    """
+    layout = ak.operations.convert.to_layout(
+        array, allow_record=False, allow_other=False
+    )
+    if isinstance(counts, int):
+        out = ak.layout.RegularArray(layout, counts)
+    else:
+        counts_layout = ak.operations.convert.to_layout(
+            counts, allow_record=False, allow_other=False
+        )
+        if isinstance(counts_layout, (
+            ak.layout.IndexedOptionArray64, ak.layout.IndexedOptionArray32
+        )):
+            mask_layout = counts_layout
+            counts_layout = mask_layout.content
+        else:
+            mask_layout = None
```

Comment:
If `counts` is an array, it needs to be a one-dimensional array of numbers, so

```python
if not type(counts).__module__.startswith("cupy."):
    counts = ak.nplike.numpy.asarray(counts)
```

will take any array-like into a NumPy or CuPy array. If `counts` is an Awkward Array, it will be (zero-copy) converted _if_ it is one-dimensional and raise an exception if it is not. This will handle any node types, not just IndexedOptionArray (and I think this handling here is not sufficiently general).

Or wait—do you have this because you want to allow and pass through missing values? That is, if there are any None values in your `counts` array, you want these to show up as None in the output? In that case, you can do

```python
if not type(counts).__module__.startswith("cupy."):
    counts = ak.operations.convert.to_numpy(counts, allow_missing=True)
    mask = ak.nplike.numpy.ma.getmask(counts)
    counts = ak.nplike.numpy.ma.filled(counts, 0)
else:
    mask = False
```

The first line in the `if` block returns a NumPy array or NumPy masked array, the second gets a boolean array of missing values or a single False if it's not a masked array, and the third line ensures that any masked values are filled with zero.

After creating the `offsets` and `out` as a ListOffsetArray64, you can then replace the `if mask_layout is not None` block with

```python
if mask is not False:
    out = ak.layout.ByteMaskedArray(ak.layout.Index8(mask.view(np.int8)), out, valid_when=False)
```

IndexedOptionArray is not the only way to make (or get) option-type data. Using `to_numpy` with `allow_missing=True` handles all of the cases, giving you a masked array, and since you have a masked array, the easiest way to use it is to make a ByteMaskedArray.

Reply:
You might also want to ensure that the `counts` array contains only integers:

```python
if not issubclass(counts.dtype.type, np.integer):
    raise ValueError(...)
```

See other examples of exception-raising in the same file for the pattern: there's a utility to ensure that all exceptions include links to this spot in GitHub.

Reply:
This is very numpy explicit. I'm also not sure about the reason behind explicitly excluding `"cupy."` as module name.
In the additional changes I made now I tried to be more general so that it should also work with cupy.

Reply:
The reason for

```python
if not type(counts).__module__.startswith("cupy."):
```

is to let only CuPy and NumPy pass through that step. If `counts` begins life as a Python sequence, Awkward Array, or anything with an `__array__` method, it will get through this block as a CuPy, NumPy, or possibly NumPy masked array (second version, which you want).

We don't yet have an idiom for this, the way we have `to_layout` to turn whatever you have into an Awkward layout node; before supporting TensorFlow, PyTorch tensors, and JAX, we'll need something like `to_nplike`, but that can be corrected later. (The fact that this advertises itself as a hardcoded check for `"cupy."` will make it stick out in a sweep—this will be one of the easy ones to find.)

Reply:
What if `counts` is an array like this
```python
ak.layout.NumpyArray(cupy.array([1,2,3]))
```
Is it proper to call `to_numpy` on this? I'm currently getting a segmentation fault when I try to do this. Do we want to transfer counts to CPU in this case?

Reply:
This was an oversight that's getting fixed in #591. What `to_numpy` is supposed to do is convert it to a NumPy array (i.e. bring it from the GPU to main memory).

The handling of GPU arrays isn't done: there are `// DERIVE` comments throughout the C++ codebase indicating spots where we're supposed to identify whether the buffers are all main memory (`kernel::lib::cpu`), all on GPUs (`kernel::lib::cuda`), or mixed (raising an exception). The Python side was mostly made NumPy/CuPy generic, but you found a missed case. I should do a more thorough sweep, like the one I'm doing for RegularArray `size == 0` in #590. We also need to get more meaningful tests for the kernels (the test cases in kernel-specification.yml haven't been checked for exhaustiveness) before implementing all the kernels in CUDA.

That's why I'm not too concerned about hard-coding the name `"cuda."` at this time: the sweep that will homogenize all of these things and make them ready for TensorFlow, PyTorch tensors, and JAX in addition to just CuPy is in the future. (That is, we're not regressing an already correct thing.)